### ISSUE TITLE
Better integration with Leela Chess Zero.

### ISSUE THREAD

**Issue by lealgo** [peterosterlund2/droidfish#47]
Hi @peterosterlund2,

I've just made this quick fork of your code here:

https://github.com/lealgo/droidfish/tree/bundle-lc0

The purpose of the fork is to include Leela Chess Zero as an internal engine. Most of the functionality is already there but I'd like you to know that any comment from you would be appreciated.

Thank you for creating this project.
Regards,
Leandro

**Comment by lealgo** [comment]
Hi again,

I've been discussing with the lc0 devs about the options for integrating the lc0 engine into different chess guis. As a first approach I made a fork of your project that includes lc0 as an internal engine. This is working fine, but we've looking for better ways. We talked about the possibility of creating an OEX package that's supported also by other guis, and we like this idea as it's more decoupled and easier to maintain. But then there's the challenge of our network weights file, AFAIK the OEX standard does not define additional dependent files, it just delivers the engine executable. As a workaround we could embed the weights file into the engine binary itself, but it's not a very elegant solution, and quite wasteful taking into account there is an executable for each architecture. Well, I just wanted to explain to you our intentions, and if you have any idea to share with us it would be much appreciated.

With best regards,
Leandro

**Comment by peterosterlund2** [comment]
Hi,

I also think that an OEX package would be better, but I don't see a good way to make such a package for Leela without extending the OEX specification somehow. It wouldn't be too hard to extend the specification but the real challenge is to get most GUIs to implement the extension. Maybe a thread in the talkchess forum would have a better chance to lead to a useful extension of the OEX specification.

Without extending the OEX specification, I think it would be possible to dynamically embed the weights file in the executable, but it would still require code in the executable to read from the embedded data instead of from an external file. I am thinking about overriding ParcelFileDescriptor openFile(Uri uri, String mode) in ChessEngineProvider, so that it first creates a temporary file consisting of the executable and the weights file concatenated, then returns a file descriptor to the temporary file.

**Comment by lealgo** [comment]
Hello,

I concur, it would be nice to extend the OEX specification, that would also allow for having opening books, etc. But I guess we're going to take the shortest path. We've got a new patch for the engine that loads the weights file from the embedded data. So I've created an OEX package containing the lc0 engine with a weights embedded, and it's working already. 

Now the size problem: We've discussed your idea of dynamic embedding, and today's patch is a step towards that as it allows the embedding to be done after the fact and at the very end of the executable, using `zip` and a plain `cat`. I believe it can be done from Java itself, overriding the openFile method as you suggest. As a stop gap measure I'm also investigating the Android's APK split feature: having gradle split the APK for each architecture, thus preventing the weights file duplication.

Thank you for your comments.
Regards,
Leandro

**Comment by lealgo** [comment]
I'd like to tell you about an issue I have while importing OEX engines on DroidFish. Using this template as a guide:

https://github.com/gkalab/chessenginesupport-androidlib/tree/master/StockfishChessEngine

I've built two packages, one unmodified with stockfish and the other with lc0. They both work well with other guis (Chess for Android and Chess PGN Master) but not with DroidFish. When I choose an OEX engine on the GUI I get an exception on `context.getContentResolver()`. I attach the exception.

![debug](https://user-images.githubusercontent.com/12534727/75847318-d48d2780-5dac-11ea-9c54-3e267da873e6.png)

![exception](https://user-images.githubusercontent.com/12534727/75847329-df47bc80-5dac-11ea-9f77-326a64fdd0c5.png)

**Comment by lealgo** [comment]
Hi,

I've tried the dynamic embedding solution and it works well, but only with DroidFish. This is a quick take:

https://github.com/lealgo/chessenginesupport-androidlib/blob/lc0/LeelaChessEngine/leelaChessEngine/src/main/java/com/kalab/chess/leelaengine/LeelaEngineProvider.java

But, I've finally found the reason why some other guis behaved differently regarding OEX. It turns out not all authors implement the OEX spec in the same way. Some like DroidFish use the ChessEngineProvider to return a file descriptor and then copy from it to a private directory. Others use the ChessEngineResolver to get the engine's full path and execute the engine directly from there. Now, the former option is extensible because it uses Android's getContentResolver to get the engine package's own "Provider" class. But the latter is not because it uses the "Resolver" class as a static library from the gui app, and the code always returns the native library as the engine fullpath. This is exactly the reason why the task killer didn't affect those other guis. And of course the same as why the debugger didn't enter such code, because the Resolver is running on the gui itself.

I'm affraid that in order to support all the guis we won't be able to use the dynamic embedding, we'll have to resort to static embedding and manage the APK size in other ways. Unless we come up with some other crazy idea. :P

Regards,
Leandro

**Comment by peterosterlund2** [comment]
The behavior was changed in the following commit (which DroidFish does not use):

https://github.com/gkalab/chessenginesupport-androidlib/commit/b16b03e372aa50a07dc71767b919715478d5f5d1

I don't think that's how content resolvers are supposed to work in Android, but if that is what other chess apps are doing I agree dynamic embedding is not a workable solution.

**Comment by lealgo** [comment]
Hi @peterosterlund2,

I'd like to inform you that today the lc0 team has released a new version of the engine that includes an Android build in the form of an OEX package. Thank you very much for your help.

https://github.com/LeelaChessZero/lc0/releases/tag/v0.24.0

With best regards,
Leandro.

**Comment by peterosterlund2** [comment]
@lealgo That is good news. The next DroidFish release will have a function to browse for files when changing UCI options (assuming a compatible file manager is installed), which should help those who want to use custom networks.

**Comment by lealgo** [comment]
Wow, that's a very nice addition! Figuring out a weights file path is the most common issue people have when they try the engine. Thank you.

I tried the new feature yesterday and it works quite well. I had to download the OI File Manager from the authors' site because for some reason I couldn't find it on Google Play. I thought about maybe using the Android's built-in file picker with the `ACTION_OPEN_DOCUMENT` intent, but then I realized it requires API level 19 that's above the min. 16 of DroidFish. Would it be possible to detect the api level and use the built-in file picker on newer devices?
